### PR TITLE
add --project-version flag for chpldoc

### LINKED ISSUE #13335: Allow setting project version number in chpldoc
Today, chpldoc generates a sphinx project which defaults to version `0.1.0`. It would be helpful for chpldoc to support something like a `--set-version` or `--project-version` flag that sets this version which gets displayed in the generated HTML.

Additionally, mason could use the mason package version to determine which flag value to pass when doing a `mason doc` call.

### PR DESCRIPTION
**Pull request by immadisairaj**
Fixes #13335 

Uses 'CHPLDOC_PROJECT_VERSION' environ variable similar to 'CHPLDOC_AUTHOR' to set the project version while using chpldoc.

It can be used like
`$ chpldoc <filename>.chpl --project-version <project-version>`
where project-version can be like 1.0.3 or 1.0.3-alpha

Adds tests for the use of 'project-version' flag

### CODE REVIEW

**Comment by lydia-duncan** [comment]
I'm going to try to get to reviewing this today, but may not manage it, there's a few meetings and I do have my own PR that I want to open.  If I haven't reviewed this in 48 hours, feel free to ping me again!

File: third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py [hunk 1]
```diff
@@ -60,9 +60,9 @@
 # built documents.
 #
 # The short X.Y version.
-version = '0.0.1'
+version = os.environ.get('CHPLDOC_PROJECT_VERSION', '')
 # The full version, including alpha/beta/rc tags.
-release = '0.0.1'
+release = os.environ.get('CHPLDOC_PROJECT_VERSION', '')
```
Comment (lydia-duncan) [inline third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py]: Based on the comments, I would expect these to do different things if an alpha/beta/rc tag was provided, but it doesn't look like they will.

Comment (immadisairaj) [reply third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py]: Yeah, I got it. Maybe I can trim the version to short at variable 'version'. That will fix it I guess.

**Comment by immadisairaj** [comment]
> How did you test this change? What cases did you check? Ideally we would track these cases in the test directory, to ensure that they don't break in the future. You'll find examples of chpldoc tests in the test/chpldoc directory - would you mind adding some tests to a new test/chpldoc/compflags/version directory? You'll probably be best served by looking in the test/chpldoc/compflags/textonly or test/chpldoc/compflags/folder directories for how to check using flags

Yeah, I will check and try to add tests.

**Comment by immadisairaj** [comment]
@lydia-duncan I suppose this is how to add the tests. Can you please review this and correct me if I am wrong. Thanks.

File: third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py (hunk 1 above)
Comment (lydia-duncan) [reply third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py]: I think you'll probably want to do some validation anyways to ensure it's in an expected version format

Comment (immadisairaj) [reply third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py]: what all type of versions can we add? maybe like '*.*.*' or '*.*.*-alpha' or '*.*.*-beta' or '*.*.*-rc'?

Comment (lydia-duncan) [reply third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py]: That seems reasonable to me, but I'd look for other Sphinx project examples online to be certain.  We don't need to support every style necessarily (at least not right away) but focusing on the most standard one seems reasonable

Comment (immadisairaj) [reply third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py]: I have added validation for the formats which I have specified earlier.

File: compiler/passes/docs.cpp [hunk 2]
```diff
@@ -295,6 +295,34 @@ std::string generateSphinxProject(std::string dirpath) {
...
+  bool check = true;
+  const char * tags[] = {"alpha", "beta", "rc"};
+  for(int i=0; i<5; i++) {
+    if(i%2 != 0 && projectVersion[i] != '.') {
+      check = false;
+      break;
+    } else if(i%2 == 0 && ((int)projectVersion[i] > (int)'9' || (int)projectVersion[i] < (int)'0')) {
+      check = false;
+      break;
+    }
```
Comment (lydia-duncan) [inline compiler/passes/docs.cpp]: This assumes that versions will always be single digits, but that's not necessarily going to be the case.  For instance, you couldn't specify `1.20.0` as the documentation version.  You also couldn't provide something that dropped the minor version, e.g `21.3`.

File: compiler/passes/docs.cpp [hunk 3]
```diff
@@ -295,6 +295,34 @@ std::string generateSphinxProject(std::string dirpath) {
...
+  }
+  if(check && projectVersion[5] == '-') {
+    if(!(strcmp(projectVersion+6,tags[0]) == 0 || strcmp(projectVersion+6,tags[1]) == 0 || strcmp(projectVersion+6,tags[2]) == 0)) {
+      check = false;
+    }
+  }
+  if(check) {
+    return projectVersion;
+  } else {
+    USR_FATAL(astr("Invalid version format: ", projectVersion));
```
Comment (lydia-duncan) [inline compiler/passes/docs.cpp]: It will be more helpful to users if you tell them why the format wasn't valid.  There's obviously a limit to what will be useful to users but not an absurd amount of effort to determine by the compiler.  In this case, maybe something like telling what character resulted in the failure?

File: compiler/passes/docs.cpp [hunk 4]
```diff
@@ -350,4 +380,4 @@ std::string filenameFromMod(ModuleSymbol *mod, std::string docsWorkDir) {
   createDocsFileFolders(filename);
 
   return filename;
-}
+}
```
Comment (lydia-duncan) [inline compiler/passes/docs.cpp]: Minor nit, this symbol means the `}` is the end of the file.  We prefer to have the file end with an empty line, so would like you to undo this change here.  It's a little difficult to explain.  As an example, take a look at the end of docs.h - you can go down once from the last declaration in this file, it doesn't end on the same line as the last semi-colon.  Does that make sense?

Comment (lydia-duncan) [reply compiler/passes/docs.cpp]: This happens a lot for your new test files

File: compiler/passes/docs.cpp [hunk 5]
```diff
@@ -295,6 +295,34 @@ std::string generateSphinxProject(std::string dirpath) {
   return std::string(moddir);
 }
 
+/* Checks for the valid Project Version (non-recursively). If an error occurs, exit and report
+ * error.
+ */
+char * checkProjectVersion(char * projectVersion) {
```
Comment (lydia-duncan) [inline compiler/passes/docs.cpp]: This function is only used in this file, so it could be declared static to the file rather than in the .h file

File: test/chpldoc/compflags/version/projectVersion.no-tag.genindex.doc.good [hunk 6]
```diff
@@ -0,0 +1,209 @@
+
+
+
+<!DOCTYPE html>
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.no-tag.genindex.doc.good]: I'm a little worried about these included files.  A lot of them have the potential to change when the sphinx version changes, meaning we'd constantly have to update them.  It would be better to only include the ones that reference the version, and add a prediff script to remove things from the .tmp file that could adjust

File: test/chpldoc/compflags/version/projectVersion.no-tag.doc.good [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+# Project: chpldoc
+# Version: 1.0.3
+# The remainder of this file is compressed using zlib.
+xÚ
Í
+Â0ïyœŠÖkßÀ Þc²4ü€PßŸK±RÅC`Í7¢ Ì7ið®Q^ÔÁtÖ«Ñ Ž@C:M°ÑÙ
ù|õè§pUgÄ

+ð#,ê+§:²w¬Khšøª7ZtYäÂcóåïðäAËgªMßÕ×è?ÂVg	Eú³
+iŽr%ÿ"zd3#¯K
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.no-tag.doc.good]: This especially is something that is likely to change frequently on us

File: test/chpldoc/compflags/version/projectVersion.doc.catfiles [hunk 8]
```diff
@@ -0,0 +1,7 @@
+docs/objects.inv                        # projectVersion.*.object.doc.good
+docs/chpl-modindex.html                 # projectVersion.*.modindex.doc.good
+docs/genindex.html                      # projectVersion.*.genindex.doc.good
+docs/index.html                         # projectVersion.*.index.doc.good
+docs/search.html                        # projectVersion.*.search.doc.good
+docs/modules/projectVersion.doc.html    # projectVersion.*.html.doc.good
+docs/_static/documentation_option.js    # projectVersion.*.option.doc.good
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.doc.catfiles]: Does specifying a .good in a .catfiles actually work?  I would not expect that to work

File: test/chpldoc/compflags/version/projectVersion.error.doc.good [hunk 9]
```diff
@@ -0,0 +1 @@
+error: Invalid version format: 1.0.3-one
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.error.doc.good]: Does this pass start_test?  I'm pretty sure start_test output always ends on the line following the last bit of text

File: test/chpldoc/compflags/version/projectVersion.no-tag.option.doc.good [hunk 10]
```diff
@@ -0,0 +1,10 @@
+var DOCUMENTATION_OPTIONS = {
+    URL_ROOT: document.getElementById("documentation_options").getAttribute('data-url_root'),
+    VERSION: '1.0.3',
+    LANGUAGE: 'None',
+    COLLAPSE_INDEX: false,
+    FILE_SUFFIX: '.html',
+    HAS_SOURCE: true,
+    SOURCELINK_SUFFIX: '.txt',
+    NAVIGATION_WITH_KEYS: false,
+};
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.no-tag.option.doc.good]: This seems like something that wouldn't need to be updated constantly, so I would lean towards continuing to check it

File: compiler/passes/docs.cpp (hunk 2 above)
Comment (immadisairaj) [reply compiler/passes/docs.cpp]: Changed it to accept also multi digits

File: compiler/passes/docs.cpp (hunk 3 above)
Comment (immadisairaj) [reply compiler/passes/docs.cpp]: Added detailed error statements also

File: compiler/passes/docs.cpp (hunk 4 above)
Comment (immadisairaj) [reply compiler/passes/docs.cpp]: reverted this change

File: compiler/passes/docs.cpp (hunk 5 above)
Comment (immadisairaj) [reply compiler/passes/docs.cpp]: made the function static

File: test/chpldoc/compflags/version/projectVersion.no-tag.option.doc.good (hunk 10 above)
Comment (immadisairaj) [reply test/chpldoc/compflags/version/projectVersion.no-tag.option.doc.good]: removed all other files and now the test checks only "docs/_static/documentation_options.js"

File: compiler/passes/docs.cpp [hunk 11]
```diff
@@ -295,6 +295,76 @@ std::string generateSphinxProject(std::string dirpath) {
...
+
+  // Supported version tags
+  const char * tags[] = {"alpha", "beta", "rc"};
+  const char * error = "";
+  for(i=0; i<length; i++) {
+    if(i>0 && projectVersion[i] == '.') {
+      if(projectVersion[i-1] != '.') {
+        dot++;
+        if(dot>2) {
+          error = "Required only two dots which seperates three numbers";
```
Comment (lydia-duncan) [inline compiler/passes/docs.cpp]: Typo: "separates", not "seperates"

Comment (lydia-duncan) [reply compiler/passes/docs.cpp]: This is not the only time that word is misspelled

File: compiler/passes/docs.cpp [hunk 12]
```diff
@@ -295,6 +295,76 @@ std::string generateSphinxProject(std::string dirpath) {
...
+      check = false;
+      break;
+    }
+  }
+  if(dot != 2 && i == length) {
+    error = "Required two dots which seperates three numbers";
+    check = false;
+  }
+  if(check && tagIndex>0) {
+    if(!(strcmp(projectVersion+tagIndex,tags[0]) == 0 || strcmp(projectVersion+tagIndex,tags[1]) == 0 || strcmp(projectVersion+tagIndex,tags[2]) == 0)) {
```
Comment (lydia-duncan) [inline compiler/passes/docs.cpp]: Minor nit, we try to keep lines less than 80 characters for legibility.  I would break this up something like this:
```C
if(!(strcmp(projectVersion+tagIndex,tags[0]) == 0 ||
   strcmp(projectVersion+tagIndex,tags[1]) == 0 ||
   strcmp(projectVersion+tagIndex,tags[2]) == 0)) {
```

File: compiler/passes/docs.cpp [hunk 13]
```diff
@@ -295,6 +295,76 @@ std::string generateSphinxProject(std::string dirpath) {
...
+        error = "Missing number before tag";
+        check = false;
+        break;
+      }
+    } else if(projectVersion[i] == '-' && dot != 2) {
+      error = "Required only two dots which seperates three numbers";
+      check = false;
+      break;
+    } else if((int)projectVersion[i] > (int)'9' || (int)projectVersion[i] < (int)'0') {
+      error = "Invalid Characters, only digits, dot and hyphen permitted";
```
Comment (lydia-duncan) [inline compiler/passes/docs.cpp]: Technically this is incorrect.  We permit other characters, but only after the hyphen and only if they match our expected tags.  I would change this error message to "... only digits and dots permitted before a hyphen" or something along those lines

File: third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py [hunk 14]
```diff
@@ -60,9 +60,10 @@
 # built documents.
 #
 # The short X.Y version.
-version = '0.0.1'
+version = os.environ.get('CHPLDOC_PROJECT_VERSION', '').split("-")[0]
+# version = '1.0.0-aplha'
```
Comment (lydia-duncan) [inline third-party/chpl-venv/chpldoc-sphinx-project/source/conf.py]: typo here, "alpha" not "aplha"

File: test/chpldoc/compflags/version/projectVersion.dotEnd.doc.good [hunk 15]
```diff
@@ -0,0 +1,11 @@
+error: Invalid version format: 21.0. due to: Cannot end with dot, can end with either number or tag
+var DOCUMENTATION_OPTIONS = {
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.dotEnd.doc.good]: It worries me a little to have the .js from a correct version included in the expected output of a failure mode.  I think you can avoid this by having all the error-generating modes run before the correct ones (or by splitting them into separate tests that write to a different output directory).

File: compiler/passes/docs.cpp (hunk 12 above)
Comment (lydia-duncan) [reply compiler/passes/docs.cpp]: It would also be reasonable to iterate through the tags array, instead of hard-coding the number of supported tags.

File: test/chpldoc/compflags/version/projectVersion.doc.chpldocopts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+--project-version 21.0.3 # projectVersion.no-tag.doc.good
+--project-version 21.0.3-alpha   # projectVersion.tag.doc.good
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.doc.chpldocopts]: It'd be nice to have a few more success modes covered by testing.  I'd like to see:
- all the supported ending tags
- a version with all single digits (e.g. `1.2.3` or whatever)
- putting the multi-digits at various places around the `.` (e.g. also covering `0.21.3`, for instance)

**Comment by immadisairaj** [comment]
I have resolved the changes as you suggested. I think now it is fine.

File: test/chpldoc/compflags/version/projectVersion.docs.cleanfiles [hunk 17]
```diff
@@ -0,0 +1 @@
+docs
```
Comment (lydia-duncan) [inline test/chpldoc/compflags/version/projectVersion.docs.cleanfiles]: Looks like this file isn't triggering because it has an extra s in the "doc" portion.

Comment (lydia-duncan) [reply test/chpldoc/compflags/version/projectVersion.docs.cleanfiles]: That causes the first five checks to fail

Comment (immadisairaj) [reply test/chpldoc/compflags/version/projectVersion.docs.cleanfiles]: Ohh! My bad. I did run the test only once and it was passing. The second time it is faling.
I'll fix it rn.

**Comment by lydia-duncan** [comment]
Okay, I got the following errors from a full test run:
```
[Error matching compiler output for compflags/thomasvandoren/chpldoc/z_copy_help.doc]
[Error matching compiler output for compflags/thomasvandoren/chpldoc/z_copyright_help_license_version.doc]
[Error matching compiler output for compflags/thomasvandoren/chpldoc/z_help_license.doc]
[Error matching compiler output for compflags/thomasvandoren/chpldoc/z_help_version.doc]
[Error matching compiler output for compflags/thomasvandoren/chpldoc.doc]
[Error matching man page(s) with compiler environment]
```

For the last one, it looks like we need to update `$CHPL_HOME/man/chpldoc.rst` for the new flag.  For the rest, the .good files in the subdirectory are all generated based on the compflags/thomasvandoren/chpldoc.doc.good file (and are not checked into the repo otherwise), so it should be sufficient to only update that file.

File: man/chpldoc.rst [hunk 18]
```diff
@@ -59,6 +59,11 @@ reStructuredText as an intermediate format.
 
     [Don't] generate HTML-based documentation (on by default)
 
+**--project-version <**\ *projectversion*\ **>**
+
+    **chpldoc** generates documentation of version *projectversion*
+    (documentation version defaults to '0.0.1' if unspecified).
```
Comment (immadisairaj) [inline man/chpldoc.rst]: Is this okay? or should I change it?

**Comment by immadisairaj** [comment]
> For the last one, it looks like we need to update `$CHPL_HOME/man/chpldoc.rst` for the new flag. 

I have updated this. Don't know if the error still occurs or not.

> For the rest, the .good files in the subdirectory are all generated based on the compflags/thomasvandoren/chpldoc.doc.good file (and are not checked into the repo otherwise), so it should be sufficient to only update that file.

These tests errors are cleared after updating the file.

**Comment by immadisairaj** [comment]
Also, I think we can update the documentation in 'CHPL_HOME/doc/html/tools/chpldoc/man.html' or will it be updated during the release?

File: man/chpldoc.rst [hunk 19]
```diff
@@ -59,6 +59,11 @@ reStructuredText as an intermediate format.
 
     [Don't] generate HTML-based documentation (on by default)
 
+**--project-version <**\ *projectversion*\ **>**
+
+    **chpldoc** generates documentation of version *projectversion*
```
Comment (lydia-duncan) [inline man/chpldoc.rst]: I prefer the text you use in docsDriver.cpp

**Comment by lydia-duncan** [comment]
> I have updated this. Don't know if the error still occurs.

Sorry, forgot that the location isn't obvious from that snippet of the output.  You can check its behavior by running start_test over the test/man/ directory

> Also, I think we can update the documentation in 'CHPL_HOME/doc/html/tools/chpldoc/man.html' or will it be updated during the release?

If I remember correctly, that file is generated by running `make docs` from $CHPL_HOME.  You can check its behavior by pointing your browser at the appropriate generated file (and I'll be sure to take a look at it as well)

File: man/chpldoc.rst (hunk 19 above)
Comment (immadisairaj) [reply man/chpldoc.rst]: Yeah sure

Comment (immadisairaj) [reply man/chpldoc.rst]: Is this good "Specify *projectversion* to set the documentation version"
or shall I proceed with "Sets documentation version to *projectversion*?

Comment (lydia-duncan) [reply man/chpldoc.rst]: I like the latter best, the first one is fine but feels more verbose for some reason

Comment (immadisairaj) [reply man/chpldoc.rst]: then I'll keep the second one and update along with the updated docs

**Comment by immadisairaj** [comment]
@lydia-duncan can you please review the changes. I hope everything is working fine and want to know if there is anything to be done.

**Comment by lydia-duncan** [comment]
Oh, sorry.  Force pushing the commits means that I don't get an email when things have changed (it also means I have to delete and repull my local copy of your branch before I can test it again, just fyi)

**Comment by immadisairaj** [comment]
> Oh, sorry. Force pushing the commits means that I don't get an email when things have changed (it also means I have to delete and repull my local copy of your branch before I can test it again, just fyi)

Does that mean in further pull requests I have to push a new commit instead of force pushing?

**Comment by lydia-duncan** [comment]
That would be helpful on my end and maintain some commit history better.  I won't necessarily enforce it, though
